tests/unit/storage/providers/LocalForageProviderTest.js

```diff
@@ -104,4 +104,26 @@ describe('storage/providers/LocalForage', () => {
                     });
             });
     });
+
+    it('clear', () => {
```

Comment:
@kidroca - @tgolen asked me to help with implementing a test like what you described [here](https://github.com/Expensify/App/issues/13884#issuecomment-1370828497) - just want to double-check check I'm capturing this right since I'm a bit new to jest and it's "fake time"...

Reply:
Almost there.
What you've got right now won't resolve the promise after 1000ms
Also `setItemTimedOut` should be a function - right now it's just the result of a `setTimeout()` call - a numeric ID

I think you just need to change the mock like so:

```js
// Given an implementation of setItem that resolves after 1000ms
localForage.setItem = jest.fn(() => new Promise(
  (resolve) =>  setTimeout(resolve, 1000))
);
```

1. `setItem` should return a promise 
2. we make the returned promise resolve after `1000ms` - it would only resolve after we advance timers

Reply:
Ah hm... in actuality, we want to measure how many times the promise is _resolved_ then, right? But there's no other way to do that other than having a callback get called when the promise resolves?

So my thought was to do something like this instead:
```JS
    it('clear', () => {
        // Use fake timers, so we can manipulate time at our will for this test.
        jest.useFakeTimers();

        // Given a mocked implementation of setItem that calls a callback function after 1000ms
        const callback = jest.fn();
        localforage.setItem = jest.fn(() => new Promise(resolve => setTimeout(resolve, 1000)).then(callback));

        // When we call setItem 5 times, but then call clear after only 1000ms
        for (let i = 0; i < 5; i++) {
            StorageProvider.setItem(`key${i}`, `value${i}`);
        }

        // We should not have called the mocked callback for setItem yet because we have not advanced the timer.
        expect(callback).not.toHaveBeenCalled();

        jest.advanceTimersByTime(1000);
        StorageProvider.clear();
        jest.advanceTimersByTime(4000);

        // The mocked callback for setItem should only have been called once since all other calls were aborted when we called clear()
        expect(callback).toHaveBeenCalledTimes(1);
    });
```

However I'm seeing that's not quite working as I would expect 🤔 
```
Error: expect(jest.fn()).toHaveBeenCalledTimes(expected)

Expected number of calls: 1
Received number of calls: 0
```

Reply:
Also, I noticed you had the comment [here](https://github.com/Expensify/react-native-onyx/blob/13eada48dc6d421f367df35fab34c001314d4e40/tests/unit/storage/providers/LocalForageProviderTest.js#L16) noting that "fake timers cause promises to hang" - wonder if that might be the same reason for the issue I'm seeing?

Reply:
(Btw, just tested this again using real timers and I don't see this problem. So I _do_ have a suspicion this is somehow related to FakeTimers)

Reply:
> Ah hm... in actuality, we want to measure how many times the promise is _resolved_ then, right?

I guess it might look like that from a certain point of view, but it's not necessary

If we make 5 `StorageProvider.setItem` calls they should be queued (this is what we test)
This should immediately call `localForage.setItem` (our mock) 1 time
then wait for the returned promise to resolve in order to move to the next item in the queue
In our test the queue would be emptied so then next 4 calls to `localForage.setItem` shouldn't happen

> Also, I noticed you had the comment [here](https://github.com/Expensify/react-native-onyx/blob/13eada48dc6d421f367df35fab34c001314d4e40/tests/unit/storage/providers/LocalForageProviderTest.js#L16) noting that "fake timers cause promises to hang" - wonder if that might be the same reason for the issue I'm seeing?

I'm can't remember what lead me to that conclusion...
Maybe we can try and update `jest` 

This might also need calls to `waitForPromisesToResolve`

> (Btw, just tested this again using real timers and I don't see this problem. So I _do_ have a suspicion this is somehow related to FakeTimers)

I'm not sure how you test this with real timers, but if you don't wait for the timers to expire the test should appear as passing successfully 

---

There's an alternative way to test this without using timers

```jsx
let i = 
const task = createDeferredTask();

localforage.setItem = jest.fn()
   . mockReturnValue(Promise.resolve()) // default value
   . mockReturnValueOnce(task.promise) // first call

for (let i = 0; i < 5; i++) {
  StorageProvider.setItem(`key${i}`, `value${i}`);
}

// at this point `localForage.setItem` should be called once, but we control when it resolves 
// and we keep it unresolved 

// this should queue any calls that follow, so we don't expect more than 1 `localForage.setItem` call
StorageProvider.clear();

return waitForPromisesToResolve().then(() => {
  expect(localForage.setItem).toHaveBeenCalledTimes(1);
})
```

Reply:
Hm, yes I do see some issues with Jest's fake timers and Promises being reported by others like [here](https://stackoverflow.com/questions/52177631/jest-timer-and-promise-dont-work-well-settimeout-and-async-function) and [here](https://stackoverflow.com/questions/51126786/jest-fake-timers-with-promises).

Although I can't quite wrap my head around your alternative way... is there a way to demonstrate the "before" case (i.e. clear() doesn't call abort and thus setItem is called 5 times?)

Reply:
> is there a way to demonstrate the "before" case (i.e. clear() doesn't call abort and thus setItem is called 5 times?)

Running this test against the pre-PR code should produce exactly that error 
expected `setItem` to be called 1 time but it was called 5 times

> I can't quite wrap my head around your alternative way

1. We create a promise which we programmatically control when to resolve
2. We configure `localforage.setItem` to return this promise the first time it's called and to otherwise return resolved promises
3. We make 5 `StorageProvider.setItem` calls
4. This adds 5 items to the queue and starts executing the first `localForage.setItem`
5. This simulates the 1st `localForage.setItem` taking a random time. 
6. The important bit is we call `StorageProvider.clear()` while the first `localForage.setItem` isn't completed yet (and the queue is not empty, because the queue would only go to the next item after the 1st completes)
7. We call `StorageProvider.clear()` and only then we resolve the first `localForage.setItem` call
8. `waitForPromisesToResolve()` makes jest wait for any promises (even promises returned as the result of a promise) to resolve
9.If `StorageProvider.clear()` does not abort the queue, more `localForage.setItem` would be executed because `StorageProvider.setItem` works like this
      - `StorageProvider.setItem` (5th call)
          - Queue -> push -> process -> 
              - return promise for 5th call
                  - return promise for 4th call
                      - return promise for 3th call
                          - return promise for nth call

So the promise of the 5th call waits the promise of the 4th call and only then calls `localForage.setItem` and the promise of the 4th call waits the promise of the 3rd and so on

Reply:
Ah, amazing explanation - thanks for taking the time to explain that to me I appreciate it! Okay, let's go with this then since it doesn't require us to update jest in this PR.